zflow_runtime/src/js/mod.rs

```diff
@@ -0,0 +1,458 @@
+use array_tool::vec::Join;
+use serde::{Deserialize, Serialize};
+use serde_json::{json, Map, Value};
+use std::marker::PhantomData;
+use std::sync::{Arc, Mutex};
+use std::{collections::HashMap, path::PathBuf};
+
+use rquickjs::{
+    module_init,
+    BuiltinLoader,
+    BuiltinResolver,
+    // loader::{
+    //     BuiltinLoader, BuiltinResolver, FileResolver, ModuleLoader, NativeLoader, ScriptLoader,
+    // },
+    Context,
+    Ctx,
+    FileResolver,
+    Func,
+    IntoJs,
+    ModuleDef,
+    ModuleLoader,
+    NativeLoader,
+    Runtime,
+};
+
+use crate::process::ProcessHandle;
+use crate::{
+    component::{Component, ComponentOptions, ModuleComponent},
+    ip::IPType,
+    port::{InPort, OutPort, PortOptions},
+    process::{ProcessError, ProcessResult},
+};
+
+#[derive(Serialize, Deserialize, Default, Clone, Debug)]
+pub struct JsComponent {
+    pub name: String,
+    pub inports: HashMap<String, PortOptions>,
+    pub outports: HashMap<String, PortOptions>,
+    #[serde(default)]
+    /// Set the default component description
+    pub description: String,
+    #[serde(default)]
+    /// Set the default component icon
+    pub icon: String,
+    #[serde(default)]
+    /// Whether the component should keep send packets
+    /// out in the order they were received
+    pub ordered: bool,
+    #[serde(default)]
+    /// Whether the component should activate when it receives packets
+    pub activate_on_input: bool,
+    #[serde(default)]
+    /// Bracket forwarding rules. By default we forward
+    pub forward_brackets: HashMap<String, Vec<String>>,
+    /// Base directory of js packages
+    pub base_dir: String,
+    /// Path to js source
+    pub source: String,
+    #[serde(default)]
+    pub package_id: String,
+    #[serde(default)]
+    pub metadata: Map<String, Value>,
+}
+
+impl JsComponent {
+    pub fn from_metadata(meta: Value) -> Option<JsComponent> {
+        JsComponent::deserialize(meta).ok()
+    }
+
+    pub fn with_metadata(&mut self, meta: Value) -> JsComponent {
+        if let Some(meta) = JsComponent::deserialize(meta.clone()).ok() {
```

Comment:
`JsComponent::from_metadata` could be used instead of JsComponent::deserialize(meta.clone()).ok(), since it does the same thing